Adds 'BASE_PATH', improved session handling, fixes #129, ...

emrysal opened the pull request:
* Contains all required changes to make Calendso run from a subfolder on a webserver of choice. Advanced hosting options HOWTO to be written, but to enable, simply set `BASE_PATH=/yourfolder` and don't forget to set `NEXTAUTH_URL`.

Combine with Nginx using the following config as inspiration:
```
location /calendso {
    proxy_set_header X-Forwarded-Host $host;
    proxy_pass http://localhost:3000;
    rewrite /calendso/(.*) /calendso/$1 break; 
}
```

**Note: router.basePath === "" if it is not applicable.**

* Fixes a bunch of missing session errors on different pages in `getServerSideProps`.
* Fixes #129 
* Simplified login, apparently this is the 'we handle it for you' approach, but you wouldn't know unless you scrolled all the way down to the bottom on https://next-auth.js.org/configuration/pages - nextauth needs some doc updates.
* Changed the target URL of the main page from `settings` to `settings/profile`, removed the Settings page, and instead turned it into a next.config.js serverside redirect.

emrysal committed:
Adds 'BASE_PATH', improved session handling, fixes #129, ...

emrysal committed:
Merge branch 'main' into feature/authentication-improvements-and-non-root-hosting

rymaju commented on pages/availability/index.tsx:
```diff
@@ -344,14 +338,17 @@ export default function Availability(props) {
                     </div>
                 }
-                <Modal heading="Start and end times changed" description="The start and end times for your day have been changed successfully." open={successModalOpen} handleClose={closeSuccessModal} />
+                <Modal heading="Start and end times changed" description="The start and end times for your day have been changed successfully." open={successModalOpen} handleClose={() => setSuccessModalOpen(false)} />
```
⛏️ I think that for the sake of consistency with the other components, we ought to keep `closeSuccessModal` and have that function `setSuccessModalOpen(false)`, since we do the same for the two other instances where we use a modal.

emrysal replied:
The two other instances are those toggle* functions which I think may lead to unintentional behaviour (what if something toggles elsewhere and then toggles again somewhere else). I think it's better to explicitly set those values as well - which would solve the consistency thing but the other way around; I suppose.

I think not having functions around modal wrappers (if no extra logic is required) is also more readable.

rymaju reviewed (commented):
Cant say much about the session stuff, looks pretty solid to me though! One nitpick I would make is to have `closeSuccessModal` still be its own function that calls `setSuccessModal(false)`.

🤔 Also just thinking out loud, it might be a good idea to abstract API calls somehow so we dont need to keep doing this `router.basePath` stuff.

emrysal commented:
I fully agree on creating a client api abstraction layer! this would also simplify handling errors that come back from the server (currently this is a bit lacking).

emrysal committed:
Removed the toggle* functions, as they actually make the code both longer and less straightforward

emrysal committed:
startTime and endTime now using dayjs, updates the day display without the need to refresh.

emrysal commented:
Realised my earlier change broke the refresh of the availability start & endTime selection, because it removed the router.path() call. So instead of re-adding that I changed it to be part of the state and also changed it to dayjs(). This removes the ability to enter out of bound numbers like the hour "2990", as it now it calculates the actual hour for 2990 (14) .

emrysal commented:
This PR addresses too much now; will split up seperate components into individual PR's for easy reviewing.
